src/main/java/com/rapid7/client/dcerpc/mslsad/messages/LsarQueryInformationPolicyRequest.java

```diff
@@ -20,29 +20,46 @@
 
 import java.io.IOException;
 import com.rapid7.client.dcerpc.io.PacketOutput;
+import com.rapid7.client.dcerpc.io.ndr.Unmarshallable;
 import com.rapid7.client.dcerpc.messages.RequestCall;
-import com.rapid7.client.dcerpc.mslsad.objects.PolicyInformationClass;
+import com.rapid7.client.dcerpc.mslsad.objects.LSAPR_POLICY_AUDIT_EVENTS_INFO;
+import com.rapid7.client.dcerpc.mslsad.objects.POLICY_INFORMATION_CLASS;
 import com.rapid7.client.dcerpc.objects.ContextHandle;
 
-public class LsarQueryInformationPolicyRequest extends RequestCall<PolicyAuditEventsInformationResponse> {
+public abstract class LsarQueryInformationPolicyRequest<T extends Unmarshallable> extends RequestCall<LsarQueryInformationPolicyResponse<T>> {
     private final static short OP_NUM = 7;
-    private final ContextHandle handle;
-    private final PolicyInformationClass infoLevel;
+    private final ContextHandle policyHandle;
+    private final POLICY_INFORMATION_CLASS infoLevel;
 
-    public LsarQueryInformationPolicyRequest(final ContextHandle handle, final PolicyInformationClass infoLevel) {
+    public LsarQueryInformationPolicyRequest(final ContextHandle policyHandle, final POLICY_INFORMATION_CLASS infoLevel) {
         super(OP_NUM);
-        this.handle = handle;
+        this.policyHandle = policyHandle;
         this.infoLevel = infoLevel;
     }
 
+    abstract T newInstance();
+
     @Override
-    public PolicyAuditEventsInformationResponse getResponseObject() {
-        return new PolicyAuditEventsInformationResponse();
+    public LsarQueryInformationPolicyResponse<T> getResponseObject() {
+        //noinspection unchecked
+        return (LsarQueryInformationPolicyResponse<T>) new LsarQueryInformationPolicyResponse(newInstance(), infoLevel);
     }
 
     @Override
     public void marshal(final PacketOutput packetOut) throws IOException {
-        packetOut.write(handle.getBytes());
-        packetOut.writeInt(infoLevel.getInfoLevel());
+        packetOut.writeMarshallable(policyHandle);
+        packetOut.writeShort(infoLevel.getInfoLevel());
+    }
+
+    public static class PolicyAuditEventsInformation extends LsarQueryInformationPolicyRequest<LSAPR_POLICY_AUDIT_EVENTS_INFO> {
```

Comment:
Can we move it to its own class given how bloated it can be with all the other information level?

Reply:
What I do like about having it in this class, is that the usage tends to look like this: `LsarQueryInformationPolicyRequest.PolicyAuditEventsInformation`.

We can move it to another class, but then we should rename `PolicyAuditEventsInformation` to something like `PolicyAuditEventsInformationRequest` at least.